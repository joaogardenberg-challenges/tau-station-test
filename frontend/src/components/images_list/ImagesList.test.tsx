import { render, screen } from '@testing-library/react'
import ThemeProvider from 'providers/Theme'

const useDispatch = jest.fn()
const useSelector = jest.fn()
const useMediaQuery = jest.fn()
const fetchImages = jest.fn()
const watchImages = jest.fn()

jest.doMock('react-redux', () => ({ useDispatch, useSelector }))
jest.doMock('@mui/material/useMediaQuery', () => useMediaQuery)
jest.doMock('actions', () => ({ fetchImages, watchImages }))

jest.mock('./ImageFiller')
jest.mock('./ImageCard')
jest.mock('./ImagesFetcher')

const ImagesList = require('./ImagesList').default

const renderWithTheme = (ui: any, options: any = {}) =>
  render(ui, { wrapper: ThemeProvider, ...options })

describe('Images List', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn())
    useSelector.mockReturnValue({ imageIds: [3, 2, 1], isFetching: false })
    useMediaQuery.mockReturnValue(true) // This means xl === true
  })

  it('renders the images successfully', () => {
    const { container } = renderWithTheme(<ImagesList />)

    const imageFillers = screen.getAllByTestId('image-filler')
    const imageCards = screen.getAllByTestId('image-card')
    const imagesFetcher = screen.getByTestId('images-fetcher')

    expect(imageFillers.length).toBe(2)
    expect(imageCards.length).toBe(3)
    expect(imagesFetcher).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('renders loading successfully', () => {
    useSelector.mockReturnValue({ imageIds: [], isFetching: true })
    const { container } = renderWithTheme(<ImagesList />)

    const spinner = screen.getByRole('progressbar')

    expect(spinner).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('renders error successfully', () => {
    useSelector.mockReturnValue({ imageIds: [], isFetching: false })
    const { container } = renderWithTheme(<ImagesList />)

    const retry = screen.getByRole('button')
    userEvent.click(retry)

    expect(stopWatchingImages).toHaveBeenCalledTimes(2)
    expect(fetchImages).toHaveBeenCalledTimes(2)
    expect(watchImages).toHaveBeenCalledTimes(2)
    expect(container).toMatchSnapshot()
  })

  it('fetches and watches the images', () => {
    renderWithTheme(<ImagesList />)

    expect(fetchImages).toHaveBeenCalledTimes(1)
    expect(watchImages).toHaveBeenCalledTimes(1)
  })
})
