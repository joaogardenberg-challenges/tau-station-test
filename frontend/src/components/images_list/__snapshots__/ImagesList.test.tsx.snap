// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Images List renders successfully 1`] = `
<div>
  <div
    class="sc-bdfBQB ecKllZ"
  >
    <ul
      class="images-list"
    >
      <div
        data-testid="image-filler"
      />
      <div
        data-testid="image-filler"
      />
      <div
        data-testid="image-card"
      />
      <div
        data-testid="image-card"
      />
      <div
        data-testid="image-card"
      />
    </ul>
    <div
      data-testid="images-fetcher"
    />
  </div>
</div>
`;
